juice/src/net/common/linear.rs

```diff
@@ -0,0 +1,263 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::co::{IBackend, ITensorDesc, SharedTensor};
+use crate::coblas::transpose::Transpose;
+use crate::net::{Context, Descriptor, Layer, LearnableParams};
+use crate::util::{native_scalar, LayerOps};
+use crate::weight::FillerType;
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct LinearConfig {
+    pub output_size: usize,
+}
+
+#[derive(Debug)]
+pub struct Linear {
+    descriptor: Descriptor,
+
+    // Weight (A) and bias (b) for the linear operation y = Ax + b.
+    weight: Rc<RefCell<LearnableParams>>,
+    bias: Rc<RefCell<LearnableParams>>,
+
+    // Constants saved for efficiency.
+    one: SharedTensor<f32>,
+    zero: SharedTensor<f32>,
+}
+
+impl LinearConfig {
+    pub fn new(output_size: usize) -> Self {
+        LinearConfig {
+            output_size: output_size,
+        }
+    }
+}
+
+impl Linear {
+    pub fn new(mut descriptor: Descriptor, config: &LinearConfig) -> Self {
+        assert_eq!(descriptor.inputs().len(), 1); // Should be only one input.
+        let input_size = descriptor.input(0).unit_shape().size();
+
+        descriptor.add_output(vec![config.output_size]);
+
+        // Create weight matrix.
+        // TODO: Explain why weights are transposed.
+        let mut weight = SharedTensor::<f32>::new(&[config.output_size, input_size]);
+        FillerType::fill_glorot(&mut weight, input_size, config.output_size);
+
+        // Create bias. Bias is typically intialized with a constant, and a suitable initialisation
+        // is stated in https://cs231n.github.io/neural-networks-2/#init for non-LSTM types.
+        let mut bias = SharedTensor::<f32>::new(&[1, config.output_size]);
+        let seed = rand::random::<f32>();
+        let bias_init_value = seed * (2.0 / seed).sqrt();
```

Comment:
This looks fishy. The normalization should be `n` not the seed according to the linked content.

Reply:
This is just copied from the existing impl: https://github.com/spearow/juice/blob/2849c96c7eeb58e4049b7055f9db08302059c443/juice/src/layers/common/linear.rs#L117.

However, re-reading the linked doc, I think the existing impl tried to follow the approach for _weights_, not bias. Specifically for bias, the doc says:

> Initializing the biases. It is possible and common to initialize the biases to be zero, since the asymmetry breaking is provided by the small random numbers in the weights. For ReLU non-linearities, some people like to use small constant value such as 0.01 for all biases because this ensures that all ReLU units fire in the beginning and therefore obtain and propagate some gradient. However, it is not clear if this provides a consistent improvement (in fact some results seem to indicate that this performs worse) and it is more common to simply use 0 bias initialization.

I've changed it to just 0.

Reply:
Thanks for cross checking and fixing it properly 🙂